sirepo/pkcli/runner.py

```diff
@@ -0,0 +1,366 @@
+                # _write_status is a no-op if there's already a status, so
+                # this really means "if we get here without having written a
+                # status, assume there was some error"
+                _write_status(runner_client.JobStatus.ERROR, run_dir)
+                # Make sure that we clear out the running job info and tell
+                # everyone the job is done, no matter what happened
+                job_info = self.jobs.pop(run_dir, None)
+                if job_info is not None:
+                    job_info.finished.set()
+
+
+_RPC_HANDLERS = {}
+
+
+def _rpc_handler(fn):
+    _RPC_HANDLERS[fn.__name__.lstrip('_')] = fn
+    return fn
+
+
+@_rpc_handler
+async def _start_job(job_tracker, request):
+    pkdc('start_job: {}', request)
+    await job_tracker.start_job(
+        request.run_dir, request.jhash, request.cmd
+    )
+    return {}
+
+
+@_rpc_handler
+async def _job_status(job_tracker, request):
+    pkdc('job_status: {}', request)
+    return {
+        'status': job_tracker.status(request.run_dir, request.jhash).value
+    }
+
+
+@_rpc_handler
+async def _cancel_job(job_tracker, request):
+    if request.run_dir not in job_tracker.jobs:
+        return {}
+    job_info = job_tracker.jobs[request.run_dir]
+    if job_info.status is not runner_client.JobStatus.RUNNING:
+        return {}
+    job_info.status = runner_client.JobStatus.CANCELED
+    _write_status(runner_client.JobStatus.CANCELED, request.run_dir)
+    job_info.process.terminate()
+    with trio.move_on_after(_KILL_TIMEOUT_SECS):
+        await job_info.finished.wait()
+    job_info.process.kill()
+    await job_info.finished.wait()
+    return {}
+
+
+# XX should we just always acquire a per-job lock here, to make sure we never
+# have to worry about different requests for the same job racing?
+async def _handle_conn(job_tracker, lock_dict, stream):
+    with _catch_and_log_errors(Exception, 'error handling request'):
+        try:
+            request_bytes = bytearray()
+            while True:
+                chunk = await stream.receive_some(_CHUNK_SIZE)
+                if not chunk:
+                    break
+                request_bytes += chunk
+            request = pkjson.load_any(request_bytes)
+            if 'run_dir' in request:
+                request.run_dir = pkio.py_path(request.run_dir)
+            pkdp('runner request: {!r}', request)
+            handler = _RPC_HANDLERS[request.action]
+            async with lock_dict[request.run_dir]:
+                response = await handler(job_tracker, request)
+            pkdp('runner response: {!r}', response)
+            response_bytes = pkjson.dump_bytes(response)
+        except Exception as exc:
+            await stream.send_all(
+                pkjson.dump_bytes({'error_string': repr(exc)})
+            )
+            # Let's also log the full error here
+            raise
+        else:
+            await stream.send_all(response_bytes)
```

Comment:
If there's a BaseException (e.g. SystemExit) it will send a response anyway. I don't think that makes sense, because we don't know that response_bytes is valid here. I think the send should be inside the try, and any BaseExceptions are simply raised implicitly.

Reply:
the `else` in `try/else` only executes if there were no exceptions, so the old behavior on `BaseException` was to send nothing at all.

But since I just removed the `except` block in response to one of your previous comments, the `try` is now pointless, and I got rid of it entirely. So now the main code and the `send_all` are all straight-line code inside the `_catch_and_log_exceptions` block.